woocommerce/woocommerce-gutenberg-products-block
# Hydrate REST endpoint data in the stores

When loading a page with a filter or products block, the new data stores do a REST API request to find all routes under the `/wc/store` namespace. This request can add a few seconds to the page load time, so we can avoid that by hydrating the data from the server.

This PR:

- Adds the route data to the block settings.
- Wraps frontend block renders with a `HydrateRestApiData` component which gets the value of the setting and passes it to the store.
- Since multiple frontend renders can occur, `HydrateRestApiData` checks for `isResolving` and `hasFinishedResolution` so the routes are only added once.

I also replaced the `forEach` with `lodash` since `lodash` is already now a frontend dependency.

### How to test the changes in this Pull Request:

1. Open the network console before using this PR on a page running the All Products block. You'll see the first API GET request goes to `/wc/store`.
2. Use this PR and load the same page. There should be no request to `/wc/store`.
3. Ensure all products continues to work as expected.

Using this same approach it may be possible to hydrate other data such as page 1 of the all products block, but for now this helps reduce the number of queries on page load.
